### DISCUSSION TITLE
HTML5 WebSockets integration supported with Ratchet

### DISCUSSION THREAD

**Discussion by wjvantilburg** [clue/framework-x#85] [Q&A] [answered]
Cool stuff, I'm hyped!

Would it be possible to receive a link or quick explanation on how to get WebSockets working?

```php
$app->get('/stream', function (ServerRequestInterface $request) {
	$stream = new ThroughStream();

	$stream->write(microtime(true) . ' hi!' . PHP_EOL);

	return new React\Http\Message\Response(
		200,
		[
			'Content-Type' => 'text/event-stream'
		],
		$stream
	);
});
```

I managed to get this from the examples and connect, but the stream is not receiving data when using WebSocket JS API. Where does Ratchet come in the picture?

~ Wouter

**Comment by clue** [comment, marked as answer]
@uuouter Thanks for bringing this up, excellent question!

WebSockets can be used by including https://github.com/ratchetphp/Ratchet, in particular by using https://github.com/voryx/WebSocketMiddleware. All packages share a common dependency tree and build on top of ReactPHP, so integration should be fairly straightforward.

As an alternative, you may also use Server-Sent Events (EventSource) which doesn't require any external dependencies, see also #68 for more details.

I'm still working on the documentation to cover this in more detail, but I hope this helps you to get started :+1:

**Comment by geertvanbommel** [comment]
A routing as simple as kemalcr.com might be cool.
```
require "kemal"

# Matches GET "http://host:port/"
get "/" do
  "Hello World!"
end

# Creates a WebSocket handler.
# Matches "ws://host:port/socket"
ws "/socket" do |socket|
  socket.send "Hello from Kemal!"
end

Kemal.run
```

**Reply by SimonFrings** [reply]
Totally agree, we also had a similar API in mind and as I also described in https://github.com/clue/framework-x/discussions/203#discussioncomment-4147061 I can see this in the nearer future for Framework X :+1:
